core/trino-main/src/main/java/io/trino/sql/planner/ConnectorExpressionTranslator.java

```diff
@@ -13,86 +13,186 @@
             this.literalEncoder = requireNonNull(literalEncoder, "literalEncoder is null");
+            this.variableMappings = requireNonNull(variableMappings, "variableMappings is null");
         }
 
-        public Expression translate(ConnectorExpression expression)
+        public Optional<Expression> process(ConnectorExpression expression)
         {
-            if (expression instanceof Variable) {
-                return variableMappings.get(((Variable) expression).getName()).toSymbolReference();
+            if (expression instanceof Constant) {
+                return Optional.of(literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType()));
+            }
+            else if (expression instanceof FieldDereference) {
+                FieldDereference fieldDereference = (FieldDereference) expression;
+                Optional<Expression> baseOpt = process(fieldDereference.getTarget());
+                return baseOpt.map(base -> new SubscriptExpression(base, new LongLiteral(Long.toString(fieldDereference.getField() + 1))));
             }
+            else if (expression instanceof Variable) {
+                String name = ((Variable) expression).getName();
+                return Optional.of(variableMappings.map(map -> map.get(name).toSymbolReference())
+                                                   .orElse(new SymbolReference(name)));
+            }
+            else if (expression instanceof Call) {
+                return processCall((Call) expression);
+            }
+            return Optional.empty();
+        }
 
-            if (expression instanceof Constant) {
-                return literalEncoder.toExpression(((Constant) expression).getValue(), expression.getType());
+        protected Optional<Expression> processCall(Call call)
+        {
+            if (LIKE_FUNCTION_NAME.equals(call.getName())) {
+                return translateLike(call.getArguments().get(0), call.getArguments().get(1));
+            }
+
+            QualifiedName qualifiedName = getQualifiedName(call);
+
+            List<Expression> arguments = call.getArguments().stream()
+                                             .map(argument -> ConnectorExpressionTranslator.translate(session, argument, metadata, literalEncoder, Optional.empty()))
+                                             .collect(Collectors.toList());
+
+            if (REGEXP_LIKE_FUNCTION_NAME.equalsIgnoreCase(call.getName()) && arguments.size() == 2) {
+                Expression pattern = arguments.get(1);
+                if (pattern instanceof StringLiteral) {
+                    Slice slice = ((StringLiteral) pattern).getSlice();
+                    if (RegexLibrary.JONI.equals(SystemSessionProperties.getRegexLibrary(session))) {
+                        JoniRegexp joniRegexp = JoniRegexpCasts.castVarcharToJoniRegexp(slice);
+                        arguments.set(1, literalEncoder.toExpression(joniRegexp, JoniRegexpType.JONI_REGEXP));
```

Comment:
This is not safe to do. Per the `Collectors.toList()` documentation:

> There are no guarantees on the type, mutability, serializability, or thread-safety of the `List` returned

Reply:
Irrelevant now that we're using `FunctionCallBuilder`